dsgrid/project.py

```diff
@@ -220,9 +201,151 @@ def _build_filtered_record_ids_by_dimension_type(self, context: QueryContext):
                     .selectExpr("from_id AS id")
                     .distinct()
                 )
+
             if dim_type in record_ids:
                 df = record_ids[dim_type].intersect(df)
+            if df.rdd.isEmpty():
+                raise DSGInvalidQuery(f"Query filter produced empty records: {dim_filter}")
             record_ids[dim_type] = df
 
-        for dimension_type, record_ids in record_ids.items():
-            context.set_record_ids_by_dimension_type(dimension_type, record_ids)
+        for dimension_type, ids in record_ids.items():
+            context.set_record_ids_by_dimension_type(dimension_type, ids)
+
+    def _process_dataset(
+        self, context: QueryContext, cached_datasets_dir: Path, dataset_id: str
+    ) -> Path:
+        """Return a Path to the created DataFrame. Does not return a DataFrame object because
+        the SparkSession will be restarted.
+
+        """
+        logger.info("Start processing query for dataset_id=%s", dataset_id)
+        project_version = f"{context.model.project.project_id}__{context.model.project.version}"
+        model_hash, text = context.model.project.dataset.params.serialize()
+        hash_dir = cached_datasets_dir / project_version / model_hash
+        if not hash_dir.exists():
+            hash_dir.mkdir(parents=True)
+            model_file = hash_dir / "model.json"
+            model_file.write_text(text)
+        cached_dataset_path = hash_dir / (dataset_id + ".parquet")
+        metadata_file = Path(str(cached_dataset_path).replace(".parquet", ".json5"))
```

Comment:
`cached_dataset_path.with_suffix(".json5")`: https://docs.python.org/3/library/pathlib.html

Reply:
Yeah, thanks. I just used that one two weeks ago elsewhere.